networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidator.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.eth2.gossip.topics.validation;
+
+import static java.lang.Math.toIntExact;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.compute_epoch_at_slot;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.compute_signing_root;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.get_domain;
+import static tech.pegasys.artemis.datastructures.util.CommitteeUtil.getAggregatorModulo;
+import static tech.pegasys.artemis.datastructures.util.CommitteeUtil.isAggregator;
+import static tech.pegasys.artemis.networking.eth2.gossip.topics.validation.ValidationResult.INVALID;
+import static tech.pegasys.artemis.util.config.Constants.DOMAIN_SELECTION_PROOF;
+import static tech.pegasys.artemis.util.config.Constants.VALID_BLOCK_SET_SIZE;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.bls.BLS;
+import tech.pegasys.artemis.bls.BLSPublicKey;
+import tech.pegasys.artemis.bls.BLSSignature;
+import tech.pegasys.artemis.datastructures.operations.AggregateAndProof;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.operations.SignedAggregateAndProof;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.CommitteeUtil;
+import tech.pegasys.artemis.datastructures.util.ValidatorsUtil;
+import tech.pegasys.artemis.storage.client.RecentChainData;
+import tech.pegasys.artemis.util.collections.ConcurrentLimitedSet;
+import tech.pegasys.artemis.util.collections.LimitStrategy;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class SignedAggregateAndProofValidator {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Set<AggregatorIndexAndSlot> receivedValidAggregations =
+      ConcurrentLimitedSet.create(VALID_BLOCK_SET_SIZE, LimitStrategy.DROP_OLDEST_ELEMENT);
+  private final AttestationValidator attestationValidator;
+  private final RecentChainData recentChainData;
+
+  public SignedAggregateAndProofValidator(
+      final AttestationValidator attestationValidator, final RecentChainData recentChainData) {
+    this.attestationValidator = attestationValidator;
+    this.recentChainData = recentChainData;
+  }
+
+  public ValidationResult validate(final SignedAggregateAndProof signedAggregate) {
```

Comment:
I don't think we're hitting this constraint:
```
The aggregate attestation defined by hash_tree_root(aggregate) has not already been seen (via aggregate gossip, within a verified block, or through the creation of an equivalent aggregate locally).
```

Reply:
No we're not.  Not really sure how to implement that to be honest. It really reaches outside the scope of the network.  Probably need to make the cache of seen attestations it's own object that can be shared around more. Added https://pegasys1.atlassian.net/browse/BC-398